Все, что нужно будет делать — запускать методы из sklearn. Вам предлагается выяснить, какое распределение лучше использовать в наивном байесовском классификаторе в зависимости от вида признаков.

Загрузите датасеты digits и breast_cancer из sklearn.datasets.

In [1]:
import sklearn.datasets

Выведите несколько строчек из обучающих выборок и посмотрите на признаки. 

In [6]:
data = sklearn.datasets.load_breast_cancer()

In [19]:
print data.feature_names, "\n", data.target_names

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension'] 
['malignant' 'benign']


In [24]:
print data.data[0-3:]

[[  1.66000000e+01   2.80800000e+01   1.08300000e+02   8.58100000e+02
    8.45500000e-02   1.02300000e-01   9.25100000e-02   5.30200000e-02
    1.59000000e-01   5.64800000e-02   4.56400000e-01   1.07500000e+00
    3.42500000e+00   4.85500000e+01   5.90300000e-03   3.73100000e-02
    4.73000000e-02   1.55700000e-02   1.31800000e-02   3.89200000e-03
    1.89800000e+01   3.41200000e+01   1.26700000e+02   1.12400000e+03
    1.13900000e-01   3.09400000e-01   3.40300000e-01   1.41800000e-01
    2.21800000e-01   7.82000000e-02]
 [  2.06000000e+01   2.93300000e+01   1.40100000e+02   1.26500000e+03
    1.17800000e-01   2.77000000e-01   3.51400000e-01   1.52000000e-01
    2.39700000e-01   7.01600000e-02   7.26000000e-01   1.59500000e+00
    5.77200000e+00   8.62200000e+01   6.52200000e-03   6.15800000e-02
    7.11700000e-02   1.66400000e-02   2.32400000e-02   6.18500000e-03
    2.57400000e+01   3.94200000e+01   1.84600000e+02   1.82100000e+03
    1.65000000e-01   8.68100000e-01   9.38700000e-01 

In [26]:
print data.target[0-3:]

[0 0 1]


In [40]:
data_digit = sklearn.datasets.load_digits()
print data_digit.target_names

[0 1 2 3 4 5 6 7 8 9]


In [41]:
print data_digit.data[0-3:]

[[  0.   0.   1.  11.  15.   1.   0.   0.   0.   0.  13.  16.   8.   2.
    1.   0.   0.   0.  16.  15.  10.  16.   5.   0.   0.   0.   8.  16.
   16.   7.   0.   0.   0.   0.   9.  16.  16.   4.   0.   0.   0.   0.
   16.  14.  16.  15.   0.   0.   0.   0.  15.  15.  15.  16.   0.   0.
    0.   0.   2.   9.  13.   6.   0.   0.]
 [  0.   0.   2.  10.   7.   0.   0.   0.   0.   0.  14.  16.  16.  15.
    1.   0.   0.   4.  16.   7.   3.  16.   7.   0.   0.   5.  16.  10.
    7.  16.   4.   0.   0.   0.   5.  14.  14.  16.   4.   0.   0.   0.
    0.   0.   0.  16.   2.   0.   0.   0.   4.   7.   7.  16.   2.   0.
    0.   0.   5.  12.  16.  12.   0.   0.]
 [  0.   0.  10.  14.   8.   1.   0.   0.   0.   2.  16.  14.   6.   1.
    0.   0.   0.   0.  15.  15.   8.  15.   0.   0.   0.   0.   5.  16.
   16.  10.   0.   0.   0.   0.  12.  15.  15.  12.   0.   0.   0.   4.
   16.   6.   4.  16.   6.   0.   0.   8.  16.  10.   8.  16.   8.   0.
    0.   1.   8.  12.  14.  12.   1.   0.]]


In [42]:
print data_digit.target[0-3:]

[8 9 8]


С помощью sklearn.cross_validation.cross_val_score c настройками по умолчанию и вызова метода mean() у возвращаемого этой функцией numpy.ndarray, сравните качество работы наивных байесовских классификаторов на этих двух датасетах. Для сравнения предлагается использовать BernoulliNB, MultinomialNB и GaussianNB. 

In [53]:
from sklearn.model_selection import cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB

cv=20
gs = GaussianNB()
gs_scores = cross_val_score(gs, data.data, data.target, cv=cv)

mn = MultinomialNB()
mn_scores = cross_val_score(mn, data.data, data.target, cv=cv)

br = BernoulliNB()
br_scores = cross_val_score(br, data.data, data.target, cv=cv)

print gs_scores.mean(), mn_scores.mean(), br_scores.mean()



0.938968710089 0.896642948367 0.627572523262


In [43]:
digit_gs_scores = cross_val_score(gs, data_digit.data, data_digit.target, cv=5)
digit_mn_scores = cross_val_score(mn, data_digit.data, data_digit.target, cv=5)
digit_br_scores = cross_val_score(br, data_digit.data, data_digit.target, cv=5)

print digit_gs_scores.mean(), digit_mn_scores.mean(), digit_br_scores.mean()


0.806520755555 0.871470302548 0.82477104598


Насколько полученные результаты согласуются с рекомендациями из лекций?

breast_cancer cогласуется с рекомендацией:
* Если в задаче есть непрерывные признаки с небольшим разбросом, то можно попробовать использовать нормальное распределение.

digits
* Если решается задача, связанная с текстами или какими-то другими разряженными дискретными при-
знаками, то хорошо подходит мультиномиальное распределение.

Для сдачи задания, ответьте на приведенные ниже вопросы.

* Вопрос 1

Каким получилось максимальное качество классификации на датасете breast_cancer?

0.938968710089

* Вопрос 2

Каким получилось максимальное качество классификации на датасете digits?

0.871470302548

* Вопрос 3

Выберите верные утверждения и запишите их номера через пробел (в порядке возрастания номера):

1) На вещественных признаках лучше всего сработал наивный байесовский классификатор с распределением Бернулли

2) На вещественных признаках лучше всего сработал наивный байесовский классификатор с мультиномиальным распределением

3) Мультиномиальное распределение лучше показало себя на выборке с целыми неотрицательными значениями признаков

4) На вещественных признаках лучше всего сработало нормальное распределение


3 4